### ISSUE TITLE
xdotool does not type keys with AltGr modifier

### ISSUE THREAD

**Issue by macdanger** [neutrinolabs/xrdp#621]
Hi guys,

I'm experimenting since some weeks with xrdp to get the best remote desktop experience. One of the problem I currently can't solve is this:
I use KeeWeb as a keepass password manager in Linux/XFCE/xrdp within a KVM. Keeweb can autotype usernames and passwords when triggered via shortcut. For this to work it uses xdotool.

Problem:
when I use Keeweb in a XFCE4 session on xrdp the '@' (at) just gets a 'q' because it seems the AltGr modifier doesn't work. When doing the exact same within a "real" X session started via lightdm everything works fine.

My current investigations shows the following difference in the input method of the started Xorg:
setxkbmap -query -v on a Xorg started via xrdp-sesman:
```
keycodes:   xfree86+aliases(qwertz)
types:      complete
compat:     complete
symbols:    pc+de(nodeadkeys)+inet(pc105)
geometry:   pc(pc105)
rules:      base
model:      pc105
layout:     de
variant:    nodeadkeys
```

When starting the same session via lightdm:
```
keycodes:   evdev+aliases(qwertz)
types:      complete
compat:     complete
symbols:    pc+de(nodeadkeys)+inet(evdev)
geometry:   pc(pc105)
rules:      evdev
model:      pc105
layout:     de
variant:    nodeadkeys

```
As you can see the "real" Xorg session uses evdev as keyboard rule while xrdp uses base. When I set 
setxkbmap -rules evdev in terminal in xrdp session then xdotool types the '@' correctly but the keyboard is not working correctly anymore, e.g. the cursor keys don't work.

I already tried to generate a new mapping file once within the "real" Xorg session and once within the xrdp session. But nothing helps.

I'm not sure if this is a problem of xdotool or Keeweb or xrdp... I can only see that it's working in a normal Xorg session started by lightdm it's not working with xrdp.

I have a feeling that this might on of several issues that arise because xrdp-sessman seems not to create a real X session like lightdm does...

Any suggestions what I could try to fix this? I guess it's a configuration issue but I don't know what else to check.

Thanks and cheers

**Comment by jsorg71** [comment]
I think this is a problem with Keeweb.  I think they don't support the base rules.
evdev is meant to be used with the evdev Linux driver.  When running X like xrdp does, base is preferred.
I did implement evdev in xrdp but there was issues.  There is a way to enable it, let me know if you want to try that.
Does Keeweb work in other environment where base is used?

**Comment by macdanger** [comment]
I'm not sure that keeweb is the problem. It seems a problem with xdotool within xrdp.
`xdotool type "test@test"` results in:
`"testqtest"` in xrdp
and 
`"test@test"` when the session is started with lightdm

But using the actual keys work and also `xdotool key "Alt_R+q"` (which is the "at" on german keyboard)

**Comment by jsorg71** [comment]
I just tested in X11rdp
xdotool type test@test and it works.
I tried both the "en-us" layout and "de"
Is this just an Xorg driver issue?
Did you try X11rdp?

```
jay@jay-All-Series:~$ setxkbmap -layout de
jay@jay-All-Series:~$ xdotool type test@test
test@testjay@jay-All-Series:~$ test@test
test@test: command not found
jay@jay-All-Series:~$ 
jay@jay-All-Series:~$ setxkbmap -layout us
jay@jay-All-Series:~$ xdotool type test@test
test@testjay@jay-All-Series:~$ test@test
test@test: command not found
jay@jay-All-Series:~$ 
jay@jay-All-Series:~$ 
jay@jay-All-Series:~$
```

**Comment by macdanger** [comment]
I will try to get the X11rdp and compile it. I'm only using Xorg drivers right now. Will come back when I tested with X11rdp...

**Comment by jsorg71** [comment]
I see this problem now with xorgxrdp(Xorg driver)

**Comment by jsorg71** [comment]
Works with X11rdp and not with xorgxrdp.
I can see the problem with xmodmap.

**Comment by macdanger** [comment]
hmmm, I compiled X11rdp and started the same XFCE session on X11rdp but had the same effect:
X11rdp:
```
keycodes:   xfree86+aliases(qwertz)
types:      complete
compat:     complete
symbols:    pc+de+inet(pc105)
geometry:   pc(pc105)
rules:      base
model:      pc105
layout:     de
```

`xdotool type "test@test"` results also in:
`testqtest`

**Comment by macdanger** [comment]
OK, after hours of reading and much trial&error I found a solution or workaround:
after reading douzends of issues with xdotool I thought: ok, well then lets try this with a normal US keyboard layout. And it worked.
But I didn't want to use US layout, instead I want to use DE. So I configured 2 layouts, de and us. This seems also to work.
My guess is that xdotool somehow always uses en layout when Xorg is configured to use base instead of evdev.
So what I did was to configure my keyboard in XFCE as having 105 keys with two layouts, the german one and additionally some random us-type keyboard. I put the german layout at the top to have it first priority. Now `setxkbmap -query -v` looks like this:
```
keycodes:   xfree86+aliases(qwertz)
types:      complete
compat:     complete
symbols:    pc+de(nodeadkeys)+inet(pc105)+us(altgr-intl):2
geometry:   pc(pc105)
rules:      base
model:      pc105
layout:     de,us
variant:    nodeadkeys,altgr-intl
```

I still don't get it why it still works with only the german layout on the "real" Xorg session but so be it :-)

I hope this will spare someone else the hours of investigating...
